packages/omgidl-serialization/src/MessageReader.ts

```diff
@@ -262,16 +278,162 @@ export class MessageReader<T = unknown> {
         } else {
           return deser(reader, arrayLengths[0]!);
         }
+      } else if (field.typeDeserInfo.type === "primitive") {
+        return field.typeDeserInfo.deserialize(
+          reader,
+          headerSpecifiedLength, // fieldLength only used for `string` type primitives
+        );
       } else {
-        const deser = deserializers.get(field.type);
-        if (deser == undefined) {
-          throw new Error(`Unrecognized primitive type ${field.type}`);
+        throw new Error(`Unhandled deserialization info type`);
+      }
+    }
+  }
+
+  /**
+   * Builds the deserialization info object for a field definition which can be a complex or primitive type.
+   *
+   * @param definition Field definition
+   * @returns Deserialization info
+   */
+  private buildFieldDeserInfo(definition: IDLMessageDefinitionField): FieldDeserializationInfo {
```

Comment:
I think this and `buildComplexDeserializationInfo` would be better suited to belong to a `DeserializationInfoCache` class outside of the `MessageReader`. This class could be instantiated with the whole definition and could hold the cache of deserializers.